:farmer: Flaky test `test_generic_pub_sub__rmw_connextdds` in linux and linux-rhel nightlies

# Description

test_generic_pub_sub__rmw_connextdds is failing in linux and linux-rhel nightlies. As it's a flaky test, it happens mostly on repeated jobs.

Reference build: 
* [rhel-debug#1967](https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_debug/1967/)
* [linux-release#3123](https://ci.ros2.org/view/nightly/job/nightly_linux_release/3123/)

Test regressions:
* [projectroot.test_generic_pub_sub__rmw_connextdds](https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_debug/1967/testReport/(root)/projectroot/test_generic_pub_sub__rmw_connextdds/)
* [test_tracetools.test_tracetools.test.test_generic_pub_sub.TestGenericPubSub.test_all](https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_debug/1967/testReport/test_tracetools.test_tracetools.test.test_generic_pub_sub/TestGenericPubSub/test_all/)

Log output (see [test report full log output](https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_debug/1967/testReport/junit/(root)/projectroot/test_generic_pub_sub__rmw_connextdds/)):
```
AssertionError: 2 != 1 : expected 1 event matching rmw_publisher_handle=24180656 in events: <Long output...>
```

Flaky ratio:
job_name|last_fail|first_fail|build_count|failure_count|failure_percentage
-- | -- | -- | -- | -- | --
nightly_linux-rhel_repeated|2024-07-02|2024-06-22|15|11|73.33
nightly_linux-rhel_release|2024-07-02|2024-06-27|15|6|40.0
nightly_linux_repeated|2024-07-01|2024-06-22|14|9|64.29
nightly_linux-rhel_debug|2024-07-01|2024-06-22|15|6|40.0
nightly_linux_release|2024-06-30|2024-06-22|15|2|13.33
nightly_linux_debug|2024-06-30|2024-06-30|15|1|6.67
Rci__nightly-connext_ubuntu_noble_amd64|2024-06-29|2024-06-22|11|6|54.55

--- a/test_tracetools/src/test_generic_ping.cpp
+++ b/test_tracetools/src/test_generic_ping.cpp
@@ -69,6 +69,7 @@ private:
     rclcpp::SerializedMessage serialized_msg;
     serialized_msg.reserve(1024);
     serializer_->serialize_message(&msg, &serialized_msg);
+    RCLCPP_INFO(this->get_logger(), "ping");
     pub_->publish(serialized_msg);
     if (do_only_one_) {
       timer_->cancel();

--- a/test_tracetools/src/test_generic_ping.cpp
+++ b/test_tracetools/src/test_generic_ping.cpp
@@ -41,7 +41,7 @@ public:
     pub_ = this->create_generic_publisher(
       PUB_TOPIC_NAME,
       "std_msgs/msg/String",
-      rclcpp::QoS(QUEUE_DEPTH));
+      rclcpp::QoS(QUEUE_DEPTH).transient_local());
     timer_ = this->create_wall_timer(
       500ms,
       std::bind(&PingNode::timer_callback, this));
@@ -70,6 +70,9 @@ private:
     serialized_msg.reserve(1024);
     serializer_->serialize_message(&msg, &serialized_msg);
     pub_->publish(serialized_msg);
+    if (do_only_one_) {
+      timer_->cancel();
+    }
   }
 
   rclcpp::GenericSubscription::SharedPtr sub_;

--- a/test_tracetools/src/test_generic_pong.cpp
+++ b/test_tracetools/src/test_generic_pong.cpp
@@ -32,7 +32,7 @@ public:
     sub_ = this->create_generic_subscription(
       SUB_TOPIC_NAME,
       "std_msgs/msg/String",
-      rclcpp::QoS(10),
+      rclcpp::QoS(10).transient_local(),
       std::bind(&PongNode::callback, this, std::placeholders::_1));
     pub_ = this->create_generic_publisher(
       PUB_TOPIC_NAME,

--- a/test_tracetools/src/test_generic_pong.cpp
+++ b/test_tracetools/src/test_generic_pong.cpp
@@ -55,6 +55,7 @@ private:
     rclcpp::SerializedMessage serialized_msg;
     serialized_msg.reserve(1024);
     serializer_->serialize_message(&next_msg, &serialized_msg);
+    RCLCPP_INFO(this->get_logger(), "pong");
     pub_->publish(serialized_msg);
     if (do_only_one_) {
       rclcpp::shutdown();

--- a/test_tracetools/src/test_ping.cpp
+++ b/test_tracetools/src/test_ping.cpp
@@ -38,7 +38,7 @@ public:
       std::bind(&PingNode::callback, this, std::placeholders::_1));
     pub_ = this->create_publisher<std_msgs::msg::String>(
       PUB_TOPIC_NAME,
-      rclcpp::QoS(QUEUE_DEPTH));
+      rclcpp::QoS(QUEUE_DEPTH).transient_local());
     timer_ = this->create_wall_timer(
       500ms,
       std::bind(&PingNode::timer_callback, this));
@@ -61,6 +61,9 @@ private:
     auto msg = std::make_shared<std_msgs::msg::String>();
     msg->data = "some random ping string";
     pub_->publish(*msg);
+    if (do_only_one_) {
+      timer_->cancel();
+    }
   }
 
   rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;

--- a/test_tracetools/src/test_ping.cpp
+++ b/test_tracetools/src/test_ping.cpp
@@ -60,6 +60,7 @@ private:
   {
     auto msg = std::make_shared<std_msgs::msg::String>();
     msg->data = "some random ping string";
+    RCLCPP_INFO(this->get_logger(), "ping");
     pub_->publish(*msg);
     if (do_only_one_) {
       timer_->cancel();

--- a/test_tracetools/src/test_pong.cpp
+++ b/test_tracetools/src/test_pong.cpp
@@ -46,6 +46,7 @@ private:
     RCLCPP_INFO(this->get_logger(), "[output] %s", msg->data.c_str());
     auto next_msg = std::make_shared<std_msgs::msg::String>();
     next_msg->data = "some random pong string";
+    RCLCPP_INFO(this->get_logger(), "pong");
     pub_->publish(*next_msg);
     if (do_only_one_) {
       rclcpp::shutdown();

--- a/test_tracetools/src/test_pong.cpp
+++ b/test_tracetools/src/test_pong.cpp
@@ -30,7 +30,7 @@ public:
   {
     sub_ = this->create_subscription<std_msgs::msg::String>(
       SUB_TOPIC_NAME,
-      rclcpp::QoS(10),
+      rclcpp::QoS(10).transient_local(),
       std::bind(&PongNode::callback, this, std::placeholders::_1));
     pub_ = this->create_publisher<std_msgs::msg::String>(
       PUB_TOPIC_NAME,

